Allow round-trip of MaskedColumn in ECSV, FITS, HDF5

taldcroft opened the pull request:
This provides a workaround for #4708.  It *allows* fully robust round-trip of `MaskedColumn` in ECSV, FITS, and HDF5 by explicitly writing the mask as a separate column (if there are any masked elements).  However, the new machinery is only enabled by default for HDF5, where writing a masked column is currently entirely unsupported.

For FITS the standard is mostly adequate, with the exception that it may not always be possible to choose a TNULL value that is not represented within the data.  In this case it would be useful to issue a warning when writing, and point to a new opt-in feature to use this MaskedColumn serialization and store the data and mask separately.

For ECSV the current standard of using `""` is robust for all cases except storing a blank entry in a string column.  Like FITS, this situation could generate a warning on write and allow for explicitly choosing to separate the data and mask.

To get MaskedColumn to work I had to introduce some generalizations in the table serialization machinery.  I added some more inline docs, though this really would deserve a longer treatment in some design docs somewhere (in a perfect world).

taldcroft commented:
BTW, test failures are numpy upstream (https://github.com/numpy/numpy/issues/11121) but I'll fix.

taldcroft commented on astropy/io/ascii/ui.py:
```diff
@@ -872,8 +872,26 @@ def write(table, output=None, format=None, Writer=None, fast_writer=True, *,
-    table = table_cls(table, names=kwargs.get('names'))
+    # Ensure that `table` is a Table subclass.
+    names = kwargs.get('names')
+    if isinstance(table, Table):
```
This has been refactored a bit but the only real diff is setting `serialized_meta` where appropriate.

taldcroft commented on astropy/table/column.py:
```diff
@@ -1033,8 +1033,58 @@ class MaskedColumnInfo(ColumnInfo):
+    def __init__(self, bound=False):
+        super().__init__(bound)
+
+        # If bound to a data object instance then create the dict of attributes
```
Similar to init code for `TimeInfo`.

taldcroft commented on astropy/table/serialize.py:
```diff
@@ -35,10 +37,13 @@ class SerializedColumn(dict):
     # If not a mixin, or if class in ``exclude_classes`` tuple then
     # treat as a normal column.  Excluded sub-classes must be explicitly
     # specified.
-    if not has_info_class(col, MixinInfo) or col.__class__ in exclude_classes:
```
This is a key change here to not use "being a mixin" as the criteria for getting into the serialize code, instead ask the object if it has columns that need to be serialized separately.

mhvk replied:
Ah, yes OOP to the rescue!

mhvk replied:
Maybe adjust the comment above this line to reflect the change.

taldcroft commented on astropy/table/serialize.py:
```diff
@@ -51,52 +56,44 @@ def _represent_mixin_as_column(col, name, new_cols, mixin_cols,
     # - meta: DO store
     info = {}
-    for attr, nontrivial, xform in (('unit', lambda x: x not in (None, ''), str),
+    for attr, nontrivial, xform in (('unit', lambda x: x is not None and x != '', str),
```
Work around bug in unit where doing an equality test of `None` with an invalid unit raises an exception.

astrofrog replied:
Please add a comment in the code for this

mhvk replied:
Better still, raise an issue!  That should be relatively trivial to fix.

taldcroft replied:
New issue #7603.

I'm not adding a comment about this because in retrospect the `x not in (None, '')` code is non-ideal anyway. It effectively expands to testing `x == None` instead of the preferred `x is None` test.

taldcroft commented on astropy/table/serialize.py:
```diff
@@ -51,52 +56,44 @@ def _represent_mixin_as_column(col, name, new_cols, mixin_cols,
 
     for data_attr in data_attrs:
         data = obj_attrs[data_attr]
-        if len(data_attrs) == 1 and not has_info_class(data, MixinInfo):
```
Replace this ad-hoc rule (and mostly repeated code) with explicit test that uses `col.info._represent_as_dict_primary_data`.

mhvk replied:
Much better indeed (even if not completely sold on profusion of names in `info` - at some point, we should have a nice, explicit API, but right now I think we're still finding out what needs to be in it!).

p.s. Having looked further down now: Do very much like the added comments.

taldcroft commented on astropy/table/serialize.py:
```diff
@@ -110,17 +107,24 @@ def _represent_mixins_as_columns(tbl, exclude_classes=()):
                                    exclude_classes=exclude_classes)
 
+    # If no metadata was created then just return the original table.
+    if not mixin_cols:
```
For code clarity this should probably be named `serialized_cols` now.

mhvk replied:
Yes, if you have the energy, that would be a good change. But perhaps easier to do in a follow-up PR, as this one is large enough already!

astrofrog commented on astropy/io/fits/tests/test_connect.py:
```diff
@@ -653,3 +654,61 @@ def test_info_attributes_with_no_mixins(tmpdir):
+    assert np.all(t2['a'] == [1, 2, 3])
+    assert np.all(t2['b'][:2] == [1.0, 2.0])  # OK
+    assert np.isnan(t2['b'][2])  # WRONG
+    assert np.all(t2['c'] == ['c', 'N', 'e'])  # WRONG ('N' is first char of 'N/A')
```
I think I'd rather have an xfail test with the right asserts than a test with the wrong asserts.

taldcroft replied:
In the end I think it makes more sense to take this "failure regression" test out and open an issue that can be tracked.  XFAILing tests is not the most sensible / useful way to track bugs.

mhvk replied:
Sounds good; just be sure there is a link to that issue in the test file.

astrofrog commented on docs/io/unified.rst:
```diff
@@ -336,6 +347,71 @@ if both a width and precision are present they are both set in the TDISPn
+      3      --    e  True
+
+  >>> for col in t.itercols():
+  ...    col.info.serialize_method['fits'] = 'data_mask'
```
Can we add a keyword argument to the Table write function for FITS files (or maybe even at a higher level so it works for HDF5 and ECSV) so that one can do:

```python
t.write('myfile.fits', serialize_method='data_mask')
```

I think this would be cleaner for users than having to loop as above.

taldcroft replied:
Agreed the interface is a little clunky and maybe we can come up with something better.  The problem is that we are dealing with different formats that have different requirements and format-specific rules and defaults.  tl;dr - it's a little complicated, and basically instead of thinking to invent something elegant I just shoe-horned onto the existing interface.

`serialize_method` came into being as a way to specify that, by default, a `Time` object should be serialized to most ASCII table formats (including ECSV) as the formatted value, but for FITS and HDF5 it should be a jd1/jd2 tuple.  The code doing the conversion would not necessarily know the context (FITS or ECSV), so we put in a context method to provide that.  So:

1. `serialize_method` defines default reasonable behaviors for column types that depend on the format (FITS, HDF5, etc).  Previously there was only `Time`, but now also `MaskedColumn`.
2. `serialize_method` exists as an Info attribute that can be customized.

Within a single table you can have `Time` and `MaskedColumn` objects, so any high-level `serialize_method` needs to allow a dict keyed by column name.  Not a problem, though if you have a masked table with Time columns you will need to fill in the dict appropriately for every column anyway.

One way forward is to open an issue to define a better high-level interface.  That is effectively independent of this PR, because no matter what you will need to retain the `Info.serialize_method` attribute for the internal code.  This provides a very convenient way to get this information down from the user interface into the low-level routines that need it.  (And of course it is already in the public API for Time).  If there is a better UI that could be what people see in the user docs.

astrofrog replied:
This makes sense. Just to clarify, we could implement this keyword argument *just* for the FITS writer though.

astrofrog commented:
I agree with your points - I think what we really need is a binary equivalent of ECSV, which at this point is basically ASDF. So we might want to recommend that if people want to save to a binary format that preserves all the Astropy-ness, the best solution is ASDF?

perrygreenfield commented:
@astrofrog You won't get an argument from me. But I do think we need other compelling examples to motivate use of asdf. I'll try to work on that in the coming months.

saimn commented on docs/io/unified.rst:
```diff
@@ -336,6 +347,71 @@ if both a width and precision are present they are both set in the TDISPn
+sentinel values according to the FITS standard:
+
+- ``NaN`` for float columns
+- Value of ``TNULLn`` for integer columns
```
Maybe mention `Column.fill_value` here ?

taldcroft replied:
Done.

saimn commented:
> For FITS the standard is mostly adequate, with the exception that it may not always be possible to choose a TNULL value that is not represented within the data. In this case it would be useful to issue a warning when writing, and point to a new opt-in feature to use this MaskedColumn serialization and store the data and mask separately.

:+1:  for adding the warning, this would close the original issue. I can have a look if you want.

taldcroft commented:
What about adding a keyword `serialize_method` to `Table.write`.  It would require a dict, which can contain keys of two different categories:

- class, such as `Time` or `MaskedColumn`.  This would apply the corresponding `serialize_method` value to all columns which are an instance of the type.
- str, This would apply `serialize_method` to the column of that name.
In the case where both apply, the `str` key version (column name) would have precedence.

So an example would be `t.write(filename, serialize_method={Time: 'jd1_jd2', 'x': 'data_mask'})`.

This would set all the appropriate `col.info.serialize_method[format] = value` in a context manager around the call to the appropriate write function.  (I haven't looked at the registry code in a while to remember if there are any stumbling blocks, but seems like it should be straightforward).

With this it would make sense to de-emphasize (or remove?) refs to the `info.serialize_method` dict.

astrofrog commented:
> What about adding a keyword serialize_method to Table.write. It would require a dict, which can contain keys of two different categories:

Sounds good but can we also make it accept a string for the case when the same serialization method can be applied to all columns?

taldcroft commented:
I've added a couple of commits that implement the suggestion to allow a `serialize_method` kwarg in `Table.write`.  In theory this PR is ready for final review from the code perspective, I just need to update the narrative docs.

taldcroft commented:
This is really ready for final review now.

mhvk commented on astropy/table/info.py:
```diff
@@ -121,3 +123,96 @@ def __call__(self, option='attributes', out=''):
+        # if an exception occurred.
+        if serialize_method:
+            for col in tbl.itercols():
+                # Go through every column and if it has a serialize_method info
```
This comment is a copy/paste artefact, I think. At least I expected it to note that this would restore the original method.

Also, why not directly iterate over `name, original_value in original_sms.items()` and do `tbl[name].info.serialize_method = original_value`?

taldcroft replied:
Good idea.  I think I just did the copy/paste without thinking.

mhvk commented on astropy/table/serialize.py:
```diff
@@ -150,6 +154,32 @@ def _construct_mixin_from_obj_attrs_and_info(obj_attrs, info):
     return mixin
 
 
+class _TableLite(OrderedDict):
```
Might it be an idea to use `TableColumns` instead? It would mean either adding a few properties to it, or slightly reworking `_construct_mixin_from_columns`.

taldcroft replied:
The methods/properties needed here are a strictly orthogonal set to `TableColumns`.  The only thing in common is the `OrderedDict` base and the concept of containing columns.  So I think it makes sense to not clutter up the core `TableColumns` class for this purpose.

mhvk replied:
OK, let's leave it for now. In the end we'd probably want `Table` to be fast enough that it doesn't matter...

taldcroft replied:
When Table is changed to allow a mixture of MaskedColumn and Column in the same table, then we can probably revert this.  Performance wasn't the driver, but the issue in the comment for this class.

mhvk commented on astropy/table/table.py:
```diff
@@ -2556,10 +2555,13 @@ def write(self, *args, **kwargs):
-        passed through to the underlying data reader (e.g. `~astropy.io.ascii.write`).
+        The arguments and keywords (other than ``format`` and
+        ``serialize_methods``) provided to this function are passed through to
+        the underlying data reader (e.g. `~astropy.io.ascii.write`).
```
Could we add a `Parameters` section here that describes `format` and `serialize_method`?

taldcroft replied:
Excellent idea, and one that I applied to `Table.read` as well.

mhvk commented on docs/io/unified.rst:
```diff
@@ -821,3 +905,86 @@ To write to a new file, the ID of the table should also be specified (unless
+and the ``mc1`` column written as a data / mask pair and
+write to ECSV:
+
+.. doctest-skip::
```
I'm surprised the doctest fails; normally it happily ignores whitespace differences.

taldcroft commented on astropy/io/ascii/ui.py:
```diff
@@ -872,8 +872,26 @@ def write(table, output=None, format=None, Writer=None, fast_writer=True, *,
+    if isinstance(table, Table):
+        # Note that making a copy of the table here is inefficient but
+        # without this copy a number of tests break (e.g. in test_fixedwidth).
+        # TODO: investigate.
```
#7605.

mhvk commented on CHANGES.rst:
```diff
@@ -59,6 +59,10 @@ astropy.io.ascii
 - Emit a warning when reading an ECSV file without specifying the ``format``
   and without PyYAML installed.  Previously this silently fell through to
   parsing as a basic format file and the file metadata was lost. [#7580]
+- Optionally allow writing masked columns to ECSV with the mask explicitly
```
Needs empty line above.

mhvk commented on astropy/table/table.py:
```diff
@@ -2527,8 +2527,25 @@ def read(cls, *args, **kwargs):
+        out : `Table`
+            Table corresponding to file contents
+
+        Notes
```
Why the notes section without actual notes? Was this meant to have something added for both `read` and `write` after the class is defined?

taldcroft replied:
Yes, I got completely confused by this.  Turns out it was some past "cleverness" [my fault!] and these docstrings get munged to add a table of available formats.  So without the "Notes" section the content gets dumped into the `Return` or `Parameters` section and the sphinx output is broken.  And if you do a `Table.read?` you'll see the table of formats.

mhvk replied:
OK, that is good to know -- but to ensure confusion doesn't recur, let's add a comment!

mhvk commented on docs/io/ascii/write.rst:
```diff
@@ -371,3 +373,76 @@ one of the specifically-allowed astropy classes.  There is no option to add
+that tells the ECSV reader to interpret the ``'c'`` and ``'c.masked'``
+columns as components of one `~astropy.table.MaskedColumn` object:
+
+.. doctest-skip::
```
Does the test really not work?

mhvk commented on docs/io/unified.rst:
```diff
@@ -336,6 +354,73 @@ if both a width and precision are present they are both set in the TDISPn
+  of later unmasking the values.
+
+Astropy provides a work-around for this limitation that users can choose to
+use.  The key part using the ``serialize_method='data_mask'`` keyword argument
```
Reword: The key part *is to use* the

mhvk commented on docs/io/unified.rst:
```diff
@@ -336,6 +354,73 @@ if both a width and precision are present they are both set in the TDISPn
+      2     2.0   -- False
+      3      --    e  True
+
+.. doctest-skip::
```
Perhaps we are skipping because this writes in the wrong place? I'm still a bit confused.

taldcroft replied:
IIRC (maybe not) some of the skipping was because PyYAML is not installed for docs generation / testing.  There was, at some point, a reason I put in the doctest-skip.  Also, in general I assume that docstring examples which demonstrate writing to a local file should be skipped.

mhvk replied:
Let me ping @astrofrog and @bsipocz - my own sense would be to have as many doctests as feasible, and it would seem OK to require yaml for them (or treat it like scipy?)

Anyway, fine to treat this as out of scope for this PR!

taldcroft replied:
And basically I don't have a good relationship with sphinx and doc testing.  For instance in that `unified.rst` file I changed a line of code within a not-skipped section to obviously fail, and my local testing (e.g. `python setup.py test --package=io --args='-k rst'`) showed no problem.  So without having a way to reliably test locally (and not having time / desire to understand our sphinx testing), I don't want to risk burning more CI time.

bsipocz replied:
It's perfectly fine to require  all optional dependencies for the docs build. However doctests runs along with normal tests, so the incantation you need most probably is `.. doctest-requires:: pyyaml` before all code snippets that requires pyyaml (or any other optional dependencies).

taldcroft replied:
Thanks @bsipocz, nice to know!  Since this is passing tests now, maybe I'll let this slide for now and clean it up later in some other context to avoid burning CI.

bsipocz replied:
Sure. That docs cleanup seems to me an ideal first PR issue, so if you could open a reminder for it we can label it accordingly to be picked up by someone else.

taldcroft commented:
@mhvk - good to merge?  I also believe that I addressed all the comments from @astrofrog .

mhvk commented:
@taldcroft - if you can, could you add/adjust the comments (see first two queries in my last review). That can be `skip ci` to not waste too much electricity.

taldcroft commented:
@bsipocz - can you (or someone) restart the ci/circleci 32bit tests?  This one never got off the ground.  The changes since last time passing are trivial, but I still feel uncomfortable merging without see at least that set of tests passing.

bsipocz commented:
@taldcroft - Done.
